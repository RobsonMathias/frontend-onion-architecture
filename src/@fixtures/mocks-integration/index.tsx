export * from './store.mock';
export * from './adapters.mock';
export * from './../mocks/http.mock';
export * from './../mocks/router.mock';
