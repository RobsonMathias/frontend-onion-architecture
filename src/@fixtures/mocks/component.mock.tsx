import React from 'react';

export function Mock() {
  return <div />;
}

export const mockComponent = (componentsName: string[] = []) => {
  let composed: { [key: string]: React.ReactNode } = {};
  componentsName.forEach((name: string) => {
    composed[name] = Mock;
  });
  return composed;
};
