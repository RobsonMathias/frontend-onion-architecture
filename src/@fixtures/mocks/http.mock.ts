export const mockHttp = {
  patch: jest.fn(),
  get: jest.fn(),
  post: jest.fn(),
  put: jest.fn(),
  delete: jest.fn(),
  interceptors: {
    request: {
      use: jest.fn(),
      eject: jest.fn()
    },
    response: {
      use: jest.fn()
    }
  }
} as any;
