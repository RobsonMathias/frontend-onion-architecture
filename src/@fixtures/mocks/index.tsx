export * from './store.mock';
export * from './router.mock';
export * from './component.mock';
export * from './wrappers';
export * from './http.mock';
export * from './services.mock';
export * from './navigation.mock';
export * from './storage.mock';
export * from './use-auth.mock';
