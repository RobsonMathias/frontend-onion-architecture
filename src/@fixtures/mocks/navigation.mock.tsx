import React from 'react';

const setData = jest.fn();
const mockUseUI = jest.fn(() => ({ setData }));

const query = { get: jest.fn(), set: jest.fn() };
const set = jest.fn();
const setAll = jest.fn();
const mockUseQuery = jest.fn(() => ({ query, setAll, set }));

const replace = jest.fn();
const mockUseURL = jest.fn(() => ({ replace }));

const mockNavBar = React.forwardRef((props, ref) => <div {...props} />);

export const mockNavigation = {
  NavBar: mockNavBar,
  useQuery: mockUseQuery,
  query,
  setAll,
  set,
  useUI: mockUseUI,
  setData,
  useURL: mockUseURL,
  replace
};
