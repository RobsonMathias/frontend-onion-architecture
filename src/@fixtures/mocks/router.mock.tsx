import React from 'react';

function Mock({ children }: any) {
  return <>{children}</>;
}

const mockUseLocation = jest.fn(() => ({}));
const goBack = jest.fn();
const push = jest.fn();
const listen = jest.fn();
const mockUseHistory = jest.fn(() => ({ push, goBack, location: {}, listen, action: 'ACTION' }));
const useRouteMatch = () => ({ push: jest.fn(), url: '/mocked' });
const mockUseParams = jest.fn();
const mockNavLink = React.forwardRef(
  ({ to, children, activeClassName, className, 'data-testid': testId }: any, ref) => (
    <a
      className={`${className}-${activeClassName}`}
      href={to}
      children={children}
      data-testid={testId}
    />
  )
);
const Link = React.forwardRef((props, ref) => <p {...props} />);
const mockRoute = React.forwardRef((props, ref) => <div />);
const mockSwitch = React.forwardRef((props, ref) => <div />);
jest.mock('react-router-dom', () => ({
  Redirect: Mock,
  BrowserRouter: Mock,
  useHistory: mockUseHistory,
  useParams: mockUseParams,
  useRouteMatch,
  NavLink: mockNavLink,
  Route: mockRoute,
  Switch: mockSwitch,
  useLocation: mockUseLocation
}));

export const mockRouter = {
  useHistory: mockUseHistory,
  Link,
  useParams: mockUseParams,
  NavLink: mockNavLink,
  useLocation: mockUseLocation,
  goBack,
  push,
  listen
};
