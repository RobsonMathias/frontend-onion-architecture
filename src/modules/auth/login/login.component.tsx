import React, { useEffect } from 'react';

import { LoginPayload } from '@types';
import { FieldErrors, useForm } from 'react-hook-form';
import { Box, Button, Grid, TextField } from '@mui/material';

export type LoginComponentProps = {
  onSubmit: (payload: LoginPayload) => void;
  loading?: boolean;
  action?: React.ReactNode;
};

export function LoginComponent(props: LoginComponentProps) {
  const { onSubmit, loading, action } = props;

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isSubmitted }
  } = useForm({ mode: 'all' });

  function hasError(error?: FieldErrors): string {
    const { message } = error || { message: '' };
    return isSubmitted ? message : '';
  }

  function submit(data: LoginPayload) {
    onSubmit(data);
  }

  useEffect(() => {
    return () => {
      reset();
    };
  }, []);

  return (
    <form onSubmit={handleSubmit(submit)}>
      <Box mb={2}>
        <TextField
          variant="filled"
          fullWidth={true}
          type="email"
          inputProps={{ 'data-testid': 'email' }}
          label="E-mail"
          error={!!hasError(errors.email)}
          helperText={hasError(errors.email)}
          placeholder="email@exemplo.com"
          {...register('email', {
            required: 'E-mail é obrigatório.'
          })}
        />
      </Box>
      <Box mb={2}>
        <TextField
          type="password"
          variant="filled"
          inputProps={{ 'data-testid': 'password' }}
          fullWidth={true}
          autoComplete="off"
          label="Senha"
          placeholder="Digite sua senha"
          error={!!hasError(errors.password)}
          helperText={hasError(errors.password)}
          {...register('password', {
            minLength: {
              value: 9,
              message: 'Senha precisa ter no mínimo 9 caracteres.'
            },
            required: 'Senha é obrigatória.'
          })}
        />
      </Box>
      <Grid container spacing={2} justifyContent="end">
        {action && <Grid item>{action}</Grid>}
        <Grid item>
          <Button
            disabled={loading}
            data-testid="submit"
            type="submit"
            variant="contained"
            color="secondary"
          >
            Acessar
          </Button>
        </Grid>
      </Grid>
    </form>
  );
}
