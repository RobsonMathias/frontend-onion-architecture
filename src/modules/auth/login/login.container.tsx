import React from 'react';

import { LoginComponent } from './login.component';
import { useLogin } from '../hooks';
import { LoginPayload } from '@types';
import { STORE_STATUS_LOADING } from 'constants/store';

export interface LoginContainerProps {
  action?: React.ReactNode;
}

export function LoginContainer(props: LoginContainerProps) {
  const { action } = props;
  const { login, auth } = useLogin();

  function handleSubmit(form: LoginPayload) {
    login(form.email, form.password);
  }

  return (
    <>
      <LoginComponent
        action={action}
        onSubmit={handleSubmit}
        loading={auth.user.status === STORE_STATUS_LOADING}
      />
    </>
  );
}
