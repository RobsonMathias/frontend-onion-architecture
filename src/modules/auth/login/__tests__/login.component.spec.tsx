/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom/extend-expect';

import { act, fireEvent, render } from '@testing-library/react';
import '@fixtures/mocks';
import React from 'react';

import { setValueForm } from './fixture';
import { LoginComponent } from '../login.component';

describe('LoginComponent', () => {
  it('should render successfully', () => {
    const onSubmit = jest.fn();
    const { container } = render(<LoginComponent onSubmit={onSubmit} />);
    expect(container).toMatchSnapshot();
  });

  it('should render action successfully', () => {
    const onSubmit = jest.fn();
    const { container } = render(
      <LoginComponent action={<span>action</span>} onSubmit={onSubmit} />
    );
    expect(container).toMatchSnapshot();
  });

  it('should submit form successfully', async () => {
    const onSubmit = jest.fn();
    const { getByTestId } = render(<LoginComponent onSubmit={onSubmit} />);

    await act(async () => {
      setValueForm(fireEvent, getByTestId);
    });

    expect(onSubmit).toHaveBeenCalledWith({ email: 'test@test.com', password: '123456789' });
  });

  it('should render required errors', async () => {
    const onSubmit = jest.fn();
    const { getByTestId, container } = render(<LoginComponent onSubmit={onSubmit} />);

    await act(async () => {
      fireEvent.click(getByTestId('submit'));
    });

    expect(container).toHaveTextContent('E-mail é obrigatório.');
    expect(container).toHaveTextContent('Senha é obrigatória.');
  });
});
