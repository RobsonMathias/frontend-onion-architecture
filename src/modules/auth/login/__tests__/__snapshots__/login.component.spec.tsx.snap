// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent should render action successfully 1`] = `
<div>
  <form>
    <div
      class="MuiBox-root css-1qm1lh"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="mui-3"
          id="mui-3-label"
        >
          E-mail
        </label>
        <div
          class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-10i04qz-MuiInputBase-root-MuiFilledInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiFilledInput-input MuiInputBase-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
            data-testid="email"
            id="mui-3"
            name="email"
            placeholder="email@exemplo.com"
            type="email"
            value=""
          />
        </div>
        
      </div>
    </div>
    <div
      class="MuiBox-root css-1qm1lh"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="mui-4"
          id="mui-4-label"
        >
          Senha
        </label>
        <div
          class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-10i04qz-MuiInputBase-root-MuiFilledInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiFilledInput-input MuiInputBase-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
            data-testid="password"
            id="mui-4"
            name="password"
            placeholder="Digite sua senha"
            type="password"
            value=""
          />
        </div>
        
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-96axuf-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <span>
          action
        </span>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-zcbmsk-MuiButtonBase-root-MuiButton-root"
          data-testid="submit"
          tabindex="0"
          type="submit"
        >
          Acessar
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`LoginComponent should render successfully 1`] = `
<div>
  <form>
    <div
      class="MuiBox-root css-1qm1lh"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="mui-1"
          id="mui-1-label"
        >
          E-mail
        </label>
        <div
          class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-10i04qz-MuiInputBase-root-MuiFilledInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiFilledInput-input MuiInputBase-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
            data-testid="email"
            id="mui-1"
            name="email"
            placeholder="email@exemplo.com"
            type="email"
            value=""
          />
        </div>
        
      </div>
    </div>
    <div
      class="MuiBox-root css-1qm1lh"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="mui-2"
          id="mui-2-label"
        >
          Senha
        </label>
        <div
          class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-10i04qz-MuiInputBase-root-MuiFilledInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiFilledInput-input MuiInputBase-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
            data-testid="password"
            id="mui-2"
            name="password"
            placeholder="Digite sua senha"
            type="password"
            value=""
          />
        </div>
        
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-96axuf-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-zcbmsk-MuiButtonBase-root-MuiButton-root"
          data-testid="submit"
          tabindex="0"
          type="submit"
        >
          Acessar
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </form>
</div>
`;
